### PR TITLE
fix: allow a teacher who has prepared a level to do something el…

### PR DESCRIPTION
**Pull request by sebp999**
…se, come back to it, and edit it further or share it

## Description
* This fixes issue #1164 which prevents a user from making a custom level, then doing something else, then opening the level and working on it again (after saving, you can only play).
* Enable editing of a specific saved level: provide a URL /rapidrouter/level_editor/<levelid> 
* Add a modal after loading saved level to choose whether to play or edit

## How Has This Been Tested?
Manually.  The changes did not break any tests as there are no Selenium tests for the level editing. I don't know if this is intentional but can make some if necessary 

## Screenshots (if appropriate):
Screenshot from 2021-03-16 15-18-35

## Checklist:

- [ x ] My change requires a change to the documentation.
- [ x ] I have updated the documentation accordingly.
- [ / ] I have linked this PR to a Zenhub Issue.

---
This change is [Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)

### CODE REVIEW

**Review by razvan-pro — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 6 unresolved discussions (waiting on @sebp999)

___
*[game/static/game/js/level_editor.js, line 151 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyQ9g-0vHD9h6dPUJY:-MWyQ9g-0vHD9h6dPUJZ:b-6bz6fu) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/static/game/js/level_editor.js#L151)):*
> ```JavaScript
>     setupToolbox();
> 
>     if (levelId != null) {
> ```

I suggest a strict check: `!==` instead of `!=`. I'm not sure if it has any effect here necessarily, but it's good practice in general unless you really need a loose check. :slightly_smiling_face: 
___
*[game/static/game/js/level_editor.js, line 3036 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyRCFgEicpC7bCDQjm:-MWyRCFgEicpC7bCDQjn:b-shlcrn) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/static/game/js/level_editor.js#L3036)):*
> ```JavaScript
>         true
>     );
>     if (LEVEL == null){
> ```

Similar here, I suggest triple equals `===` instead :slightly_smiling_face: 
___
*[game/templates/game/level_editor.html, line 23 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyTzZ57FfvsO3xz133:-MWyTzZ57FfvsO3xz134:b-i18pi9) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/templates/game/level_editor.html#L23)):*
> ```HTML
>     {% else %}
>     var LEVEL = null;
>     {% endif %}
> ```

Would this work and look better/be easier to read? :slightly_smiling_face: `var LEVEL = {{ level if level is not None else 'null' }};`
___
*[game/views/level_editor.py, line 95 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyUhFV90tNcAzphNZh:-MWyUhFV90tNcAzphNZi:b-dbfupc) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L95)):*
> ```Python
>     }
>     if levelId:
>         context["level"] = levelId
> ```

Can we get rid of the `if` and set `"level"` to `levelId` directly in the context dictionary above? I assume if it will be `None` it will just be passed to the template and we can do the check there.
___
*[game/views/level_editor.py, line 246 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWy4-zl-FNjZb3ZCC-w:-MWy4-zl-FNjZb3ZCC-x:bk9k7z9) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L246)):*
> ```Python
>     data = json.loads(request.POST["data"])
>     if not data["character"]:
>         print("add character")
> ```

Is this `print` needed? I would suggest `logging` if so, but I see it's not being set up/used anywhere in the project :grin: 
___
*[game/views/level_editor.py, line 247 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyVEaJ3OOsrnq0OZZp:-MWyVEaJ3OOsrnq0OZZq:b-64izt) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L247)):*
> ```Python
>     if not data["character"]:
>         print("add character")
>         data["character"] = 1
> ```

Can we add a test for this `if` in `test_level_editor.py`? This would be the only line not covered if we remove the `if levelId` above and should help pass the coverage check :heavy_check_mark:

**Review by faucomte97 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 7 unresolved discussions (waiting on @sebp999)

___
*[game/urls.py, line 165 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MXbPGog7ZjZoJ4VK9uG:-MXbPGog7ZjZoJ4VK9uH:bm9vgew) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/urls.py#L165)):*
> ```Python
>                 url(r"^$", level_editor, name="level_editor"),
>                 url(
>                     r"^(?P<levelId>[0-9]+)/$",
> ```

what happens if we input a URL which is a random number here, will it open up a level that we do not own, or does it check for permissions before doing so?

File: Pipfile [hunk 1]
```diff
@@ -4,8 +4,8 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-rapid-router = {path = ".",editable = true}
 django-autoconfig = { git = 'https://github.com/ocadotechnology/django-autoconfig.git', ref = 'master' }
+rapid-router = {editable = true, path = "."}
```
Comment (sebp999) [inline Pipfile:8]: This is because pipenv installs from the repo instead of as editable.  I can't find it as a bug, but this seems to work.

**Comment by sebp999** [comment]
Pushed another commit to deal with the above

**Review by razvan-pro — COMMENTED** [review]
> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 4 unresolved discussions (waiting on @sebp999)

___
*[game/templates/game/level_editor.html, line 24 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWeUwrr6GNXATeCVXjN-r2-24:-MYJkCdn1cpiVXTbyGmN:b-3uht1k) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/game/templates/game/level_editor.html#L24)):*
<details><summary><i>Previously, sebp999 wrote…</i></summary><blockquote>

You can't have a list comprehension in a template :(
</blockquote></details>

Oh, I was referring to an inline-if :slightly_smiling_face: This seems to be possible? https://jinja.palletsprojects.com/en/2.11.x/templates/#if-expression
___
*[game/views/level_editor.py, line 95 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyUhFV90tNcAzphNZh:-MYJxJn-1Nc_Msao1YIV:b-mp02f9) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L95)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Can we get rid of the `if` and set `"level"` to `levelId` directly in the context dictionary above? I assume if it will be `None` it will just be passed to the template and we can do the check there.
</blockquote></details>

Oh, it needs extra checks I see - looks good :+1:

**Review by faucomte97 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 4 unresolved discussions (waiting on @razvan-pro and @sebp999)

___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZCkv6I5MAd5OY_FJMp:b-2hq00b) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, sebp999 wrote…</i></summary><blockquote>

This is because pipenv installs from the repo instead of as editable.  I can't find it as a bug, but this seems to work.
</blockquote></details>

@razvan-pro that's interesting, we probably need to do that for the other repos
___
*[game/urls.py, line 165 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MXbPGog7ZjZoJ4VK9uG:-MZCkgZ31XVcusxkv_nL:b-7aw3p3) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/urls.py#L165)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

what happens if we input a URL which is a random number here, will it open up a level that we do not own, or does it check for permissions before doing so?
</blockquote></details>

@sebp999 just wanted to check if this comment was OK to resolve?

**Comment by sebp999** [comment]
> > _[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)_ status: all files reviewed, 7 unresolved discussions (waiting on @sebp999)
> 
> _[game/urls.py, line 165 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MXbPGog7ZjZoJ4VK9uG:-MXbPGog7ZjZoJ4VK9uH:bm9vgew) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/urls.py#L165)):_
> 
> > ```python
> >                 url(r"^$", level_editor, name="level_editor"),
> >                 url(
> >                     r"^(?P<levelId>[0-9]+)/$",
> > ```
> 
> what happens if we input a URL which is a random number here, will it open up a level that we do not own, or does it check for permissions before doing so?

Yes, it checks: 

> > _[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)_ status: all files reviewed, 7 unresolved discussions (waiting on @sebp999)
> 
> _[game/urls.py, line 165 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MXbPGog7ZjZoJ4VK9uG:-MXbPGog7ZjZoJ4VK9uH:bm9vgew) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/urls.py#L165)):_
> 
> > ```python
> >                 url(r"^$", level_editor, name="level_editor"),
> >                 url(
> >                     r"^(?P<levelId>[0-9]+)/$",
> > ```
> 
> what happens if we input a URL which is a random number here, will it open up a level that we do not own, or does it check for permissions before doing so?

Yes, it checks: level_editor.py line 94 and test_level_editor.py:246

**Comment by sebp999** [comment]
Thanks, yes, it does check for ownership: it's level_editor.py:94 and test_level_editor:246

**Review by razvan-pro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 4 unresolved discussions (waiting on @faucomte97 and @sebp999)

___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZDOnTh4ZynZN_X63Yg:btny1va) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

@razvan-pro that's interesting, we probably need to do that for the other repos
</blockquote></details>

I don't really get this change - does this actually change anything? I thought it's equivalent to the previous one :thinking: Does this install it as editable while the previous installed it from the repo?

**Review by faucomte97 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 3 unresolved discussions (waiting on @sebp999)

___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZDQdr9D7L1cRLNk-4t:b-t3iii) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

I don't really get this change - does this actually change anything? I thought it's equivalent to the previous one :thinking: Does this install it as editable while the previous installed it from the repo?
</blockquote></details>

@razvan-pro check out @sebp999 's comment just above :stuck_out_tongue:

**Review by faucomte97 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-:-MZDQmvo6saBuAmofeBd:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @sebp999)

**Review by razvan-pro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 3 unresolved discussions (waiting on @sebp999)

___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZIfgP98SR-UZOgGBPb:b-2m0or4) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

@razvan-pro check out @sebp999 's comment just above :stuck_out_tongue: 
</blockquote></details>

@faucomte97 I read the comment, but I still don't understand, sorry, maybe I'm missing some context :sweat_smile: The change looks fine to me though, but I'm not sure if the order change would make any difference :shrug:

**Comment by sebp999** [comment]
___
*[game/templates/game/level_editor.html, line 24 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWeUwrr6GNXATeCVXjN-r2-24:-MYKVTmGBW-RvJFr0K9Q:b-cno4gi) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/game/templates/game/level_editor.html#L24)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Oh, I was referring to an inline-if :slightly_smiling_face: This seems to be possible? https://jinja.palletsprojects.com/en/2.11.x/templates/#if-expression
</blockquote></details>

I don't know why I said list comprehension, I'm sorry!  I meant to say "I don't think you can have inline logic".  
I tried it but get a `TemplateSyntaxError` 
```In template /home/seb/dev/rapid-router/game/templates/game/level_editor.html, error at line 19

Could not parse the remainder: ' if level is not None else 'null'' from 'level if level is not None else 'null''
 ```
so I guess the project can't parse Jinja templates

**Comment by sebp999** [comment]
___
*[game/static/game/js/level_editor.js, line 3036 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyRCFgEicpC7bCDQjm:-MZIv1NL5SY1NMB81NHM:bb6un1j) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/static/game/js/level_editor.js#L3036)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Similar here, I suggest triple equals `===` instead :slightly_smiling_face: 
</blockquote></details>

Done

**Comment by sebp999** [comment]
___
*[game/static/game/js/level_editor.js, line 151 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyQ9g-0vHD9h6dPUJY:-MZIusDa7mEEAxaC90zN:bb6un1j) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/static/game/js/level_editor.js#L151)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

I suggest a strict check: `!==` instead of `!=`. I'm not sure if it has any effect here necessarily, but it's good practice in general unless you really need a loose check. :slightly_smiling_face: 
</blockquote></details>

Done

**Comment by sebp999** [comment]
___
*[game/views/level_editor.py, line 246 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWy4-zl-FNjZb3ZCC-w:-MZIvYn89Erf-J5dNy_M:b-9pe4u) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L246)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Is this `print` needed? I would suggest `logging` if so, but I see it's not being set up/used anywhere in the project :grin: 
</blockquote></details>

It's removed now

**Comment by sebp999** [comment]
___
*[game/views/level_editor.py, line 247 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyVEaJ3OOsrnq0OZZp:-MZIvhSy2d0JM_MuOEo2:b-1575zy) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/views/level_editor.py#L247)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Can we add a test for this `if` in `test_level_editor.py`? This would be the only line not covered if we remove the `if levelId` above and should help pass the coverage check :heavy_check_mark:  
</blockquote></details>

There is a test for this now in test_level_editor.py l 246

**Comment by sebp999** [comment]
___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZIw-3jA-duZj7uUVLN:bpi2jwy) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

@faucomte97 I read the comment, but I still don't understand, sorry, maybe I'm missing some context :sweat_smile: The change looks fine to me though, but I'm not sure if the order change would make any difference :shrug: 
</blockquote></details>

I noticed that in my machine it was installing the package as a package, not editable.  Doing it this way round seems to make it editable.  I don't know why.  Leave it if it works for you and I'll uninstall it when I get that!

**Review by razvan-pro — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-:-MZJ-7MAAkpPr2XpztA1:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @faucomte97, @razvan-pro, and @sebp999)

___
*[game/templates/game/level_editor.html, line 24 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWeUwrr6GNXATeCVXjN-r2-24:-MZIzw8G1kMyJ_WJGWQT:b7jvnsa) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/game/templates/game/level_editor.html#L24)):*
<details><summary><i>Previously, sebp999 wrote…</i></summary><blockquote>

I don't know why I said list comprehension, I'm sorry!  I meant to say "I don't think you can have inline logic".  
I tried it but get a `TemplateSyntaxError` 
```In template /home/seb/dev/rapid-router/game/templates/game/level_editor.html, error at line 19

Could not parse the remainder: ' if level is not None else 'null'' from 'level if level is not None else 'null''
 ```
so I guess the project can't parse Jinja templates




</blockquote></details>

No worries! Ah, that's weird :thinking: But it's alright, we can keep it like this then :slightly_smiling_face:

**Review by razvan-pro — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: all files reviewed, 2 unresolved discussions (waiting on @sebp999)

___
*[Pipfile, line 8 at r2](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MY4JdDk-hsiop-qPsPJ-r2-8:-MZJ-FIY9dhuisNqxqr6:brfremf) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/bf133a8e98d37ad04cec737bc77902a680d5b60d/Pipfile#L8)):*
<details><summary><i>Previously, sebp999 wrote…</i></summary><blockquote>

I noticed that in my machine it was installing the package as a package, not editable.  Doing it this way round seems to make it editable.  I don't know why.  Leave it if it works for you and I'll uninstall it when I get that!
</blockquote></details>

No problem, I'm happy with this! :+1:

**Comment by sebp999** [comment]
___
*[game/templates/game/level_editor.html, line 23 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170#-MWyTzZ57FfvsO3xz133:-M_zKub33MRxiwETzjta:be7ckoc) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/0e94ef9c772c933c869cd688875bb848b0b65f9b/game/templates/game/level_editor.html#L23)):*
<details><summary><i>Previously, razvan-pro (Razvan Mahu) wrote…</i></summary><blockquote>

Would this work and look better/be easier to read? :slightly_smiling_face: `var LEVEL = {{ level if level is not None else 'null' }};`
</blockquote></details>

Sadly it doesn't work in Django template :(

**Review by razvan-pro — APPROVED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sebp999)

**Review by faucomte97 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1170)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sebp999)

**Pull request merged**